rs/nervous_system/integration_tests/tests/upgrade_sns_controlled_canister_with_large_wasm.rs

```diff
@@ -0,0 +1,218 @@
+    // 1. Prepare the world
+    let pocket_ic = PocketIcBuilder::new()
+        .with_nns_subnet()
+        .with_sns_subnet()
+        .with_application_subnet()
+        .build_async()
+        .await;
+
+    // Install the NNS canisters.
+    {
+        let with_mainnet_nns_canisters = false;
+        install_nns_canisters(&pocket_ic, vec![], with_mainnet_nns_canisters, None, vec![]).await;
+    }
+
+    // Publish SNS Wasms to SNS-W.
+    {
+        let with_mainnet_sns_canisters = false;
+        add_wasms_to_sns_wasm(&pocket_ic, with_mainnet_sns_canisters)
+            .await
+            .unwrap();
+    };
+
+    // Install a dapp canister.
+    let original_wasm = Wasm::from_bytes(UNIVERSAL_CANISTER_WASM.to_vec());
+    let original_wasm_hash = original_wasm.sha256_hash();
+
+    let app_subnet = pocket_ic.topology().await.get_app_subnets()[0];
+
+    let target_canister_id = install_canister_on_subnet(
+        &pocket_ic,
+        app_subnet,
+        vec![],
+        Some(original_wasm.clone()),
+        vec![ROOT_CANISTER_ID.into()],
+    )
+    .await;
+
+    let sns = {
+        let create_service_nervous_system = CreateServiceNervousSystemBuilder::default()
+            .with_dapp_canisters(vec![target_canister_id])
+            .build();
+
+        let swap_parameters = create_service_nervous_system
+            .swap_parameters
+            .clone()
+            .unwrap();
+
+        let sns_instance_label = "1";
+        let (sns, _) = nns::governance::propose_to_deploy_sns_and_wait(
+            &pocket_ic,
+            create_service_nervous_system,
+            sns_instance_label,
+        )
+        .await;
+
+        sns::swap::await_swap_lifecycle(&pocket_ic, sns.swap.canister_id, Lifecycle::Open)
+            .await
+            .unwrap();
+        sns::swap::smoke_test_participate_and_finalize(
+            &pocket_ic,
+            sns.swap.canister_id,
+            swap_parameters,
+        )
+        .await;
+
+        sns
+    };
+
+    let store_canister_id = if store_same_as_target {
+        target_canister_id
+    } else {
+        install_canister_on_subnet(
+            &pocket_ic,
+            app_subnet,
+            vec![],
+            None,
+            vec![sns.root.canister_id],
+        )
+        .await
+    };
+
+    // TODO: Make the new WASM bigger than 2 MiB so that it does not fit into an ingress message.
+    let new_wasm = modify_wasm_bytes(&original_wasm.bytes(), 123);
+    let new_wasm = Wasm::from_bytes(new_wasm.to_vec());
+    let new_wasm_hash = new_wasm.sha256_hash();
+
+    // Smoke test
+    assert_ne!(new_wasm_hash, original_wasm_hash);
+
+    let chunk_hashes_list = {
+        pocket_ic
+            .upload_chunk(
+                store_canister_id.into(),
+                // This is a simplification; for now, we assume the Root itself decides to upload
+                // some WASM chunks, but eventually this should be triggered via proposal.
```

Comment:
Actually, it doesn't matter how the chunks get there so long as 1) root is a co-controller and 2) the wasm hash is correct (which is part of the proposal and should be verified).  I think that means that you don't need root to upload the chunks necessarily.